x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/StreamingHttpResultPublisher.java

```diff
@@ -0,0 +1,218 @@
+
+            // always check if totalByteSize > the configured setting in case the settings change
+            if (totalByteSize.accumulateAndGet(allBytes.length, Long::sum) >= settings.getMaxResponseSize().getBytes()) {
+                pauseProducer(ioControl);
+            }
+
+            // always run in case we're waking up from a pause and need to start a new thread
+            runOrScheduleNextMessage();
+        }
+    }
+
+    private void pauseProducer(IOControl ioControl) {
+        ioControl.suspendInput();
+        synchronized (ioLock) {
+            savedIoControl = ioControl;
+        }
+    }
+
+    private void resumeProducer() {
+        synchronized (ioLock) {
+            if (savedIoControl != null) {
+                savedIoControl.requestInput();
+                savedIoControl = null;
+            }
+        }
+    }
+
+    // if there is currently a thread running in a loop, it should pick up this new request
+    // if not, check if this thread is one of ours and reuse it
+    // else, offload to a new thread so we do not block Apache's connection thread or ElasticSearch's transport thread
+    private void runOrScheduleNextMessage() {
+        if (subscriber != null && isRunning.compareAndSet(false, true)) {
+            var currentThreadPool = EsExecutors.executorName(Thread.currentThread().getName());
+            if (UTILITY_THREAD_POOL_NAME.equalsIgnoreCase(currentThreadPool)) {
+                new OffloadThread().run();
+            } else {
+                threadPool.executor(UTILITY_THREAD_POOL_NAME).execute(new OffloadThread());
+            }
+        }
+    }
+
+    @Override
+    public void responseCompleted(HttpContext httpContext) {}
+
+    @Override
+    public void failed(Exception e) {
+        if (this.isDone.compareAndSet(false, true)) {
+            this.ex = e;
+        }
+        if (isRunning.get() == false && subscriber != null && pendingRequest.compareAndSet(true, false)) {
+            subscriber.onError(ex);
+        }
+    }
+
+    @Override
+    public Exception getException() {
+        return ex;
+    }
+
+    @Override
+    public HttpResponse getResult() {
+        return response;
+    }
+
+    @Override
+    public boolean isDone() {
+        return isDone.get();
+    }
+
+    @Override
+    public void close() {
+        // if isRunning == true, the OffloadThread will close out the subscription
+        // if pendingRequest is false, the next call to request() will close out the subscription
+        if (isDone.compareAndSet(false, true)
+            && isRunning.get() == false
+            && subscriber != null
+            && pendingRequest.compareAndSet(true, false)) {
+            subscriber.onComplete();
+        }
+    }
+
+    @Override
+    public boolean cancel() {
+        resultQueue.clear();
+        close();
+        return true;
+    }
+
+    private class HttpSubscription implements Flow.Subscription {
+        @Override
+        public void request(long n) {
```

Comment:
Should the code respect the `n` parameter? If the caller has asked for up to `n` responses and it gets more will that be a problem?